tests/test_kb_auth.py

```diff
@@ -0,0 +1,123 @@
+
+config = bootstrap_config()
+AUTH_URL = config["staging_service"]["AUTH_URL"]
+
+# NOTE: These tests are intended to be run against the auth service at
+# https://ci.kbase.us/services/auth
+# Any unexpected failures should check the auth service, the given test
+# token, and whether the test user exists or not.
+
+TEST_TOKEN = os.environ.get("KB_TEST_TOKEN")
+TEST_USER = os.environ.get("KB_TEST_USER")
+VALID_TEST_USER = "narrativetest"
+NOT_REAL_USER = "please_do_not_ever_make_this_username_what_is_wrong_with_you"
+INVALID_USER = "_(__)_"
+
+SKIP_AUTH_TESTS = not TEST_TOKEN or not TEST_USER
+
+
+@pytest.fixture
+async def auth_client():
+    """
+    A fixture that makes a default auth client with a shiny new cache.
+    Avoids the extra lag from using KBaseAuth.create.
+    """
+    return await KBaseAuth.create(AUTH_URL)
+
+
+async def test_create():
+    auth_client = await KBaseAuth.create(AUTH_URL)
+    assert isinstance(auth_client, KBaseAuth)
+
+
+async def test_create_non_auth_url():
+    with pytest.raises(IOError, match="does not appear to be the KBase Authentication Service"):
+        await KBaseAuth.create("https://github.com")
+
+
+async def test_create_incorrect_url():
+    with pytest.raises(IOError, match="Error from KBase auth server"):
+        await KBaseAuth.create(AUTH_URL + "wrong")
+
+
+async def test_non_json_error():
+    with pytest.raises(IOError, match="Non-JSON response from KBase auth server"):
+        await KBaseAuth.create("https://kbase.us/services")
+
+
+# TODO: test other service fail modes
+@pytest.mark.skipif(SKIP_AUTH_TESTS, reason="test auth credentials not set")
+async def test_get_user_from_cache(auth_client: KBaseAuth):
+    # patch this to act as a spy, should only be called once with the caching done.
+    with patch("staging_service.kb_auth_client._get", wraps=_get) as spy_get:
+        result = await auth_client.get_user(TEST_TOKEN)
+        assert result == TEST_USER
+
+        result = await auth_client.get_user(TEST_TOKEN)
+        assert result == TEST_USER
+
+        # make sure it's only called once
+        spy_get.assert_awaited_once()
+
+
+@pytest.mark.skipif(SKIP_AUTH_TESTS, reason="test auth credentials not set")
+async def test_get_user_drop_cache():
+    auth_client = await KBaseAuth.create(AUTH_URL, cache_expiration=1)
+    with patch("staging_service.kb_auth_client._get", wraps=_get) as spy_get:
+        result = await auth_client.get_user(TEST_TOKEN)
+        assert result == TEST_USER
+
+        result = await auth_client.get_user(TEST_TOKEN)
+        assert result == TEST_USER
+
+        # make sure it's only called once at this point
+        spy_get.assert_awaited_once()
+        await asyncio.sleep(1)
+
+        # cache should have dropped the user
+        result = await auth_client.get_user(TEST_TOKEN)
+        assert result == TEST_USER
+
+        # expect this to be 2
+        assert spy_get.call_count == 2
+
+
+# TODO: test cache max size?
+
```

Comment:
It seems like you never test get_user success without mocking out the auth call unless I'm missing something

Reply:
Auth call's not mocked, it's wrapped (line 72 `wraps=_get`). `_get` still runs and returns as usual, this just effectively lets the test track how many times it was called. Should be exactly 2, total, so 1 call uses the cache.

I can throw a simple get_user success call if you prefer.